tests/test_templatetags.py

```diff
@@ -2392,3 +2392,215 @@ def test_inner_slot_iteration_nested_with_slot_default_and_outer_scope_variable_
+                def
+                456
+            </div>
+        """
+        self.assertHTMLEqual(rendered, expected)
+
+    def test_slot_data_raises_on_slot_data_and_as_same_var(self):
+        @component.register("test")
+        class TestComponent(component.Component):
+            template: types.django_html = """
+                {% load component_tags %}
+                <div>
+                    {% slot "my_slot" abc=abc 123=var123 %}Default text{% endslot %}
+                </div>
+            """
+
+            def get_context_data(self):
+                return {
+                    "abc": "def",
+                    "var123": 456,
+                }
+
+        template: types.django_html = """
+            {% load component_tags %}
+            {% component "test" %}
+                {% fill "my_slot" slot_data="slot_var" as "slot_var" %}
+                    {{ slot_var.default }}
+                {% endfill %}
+            {% endcomponent %}
+        """
+        with self.assertRaisesMessage(
+            TemplateSyntaxError,
+            "'fill' received the same string for slot alias (as ...) and slot data (slot_data=...)",
+        ):
+            Template(template).render(Context())
+
+    def test_slot_data_fill_without_data(self):
+        @component.register("test")
+        class TestComponent(component.Component):
+            template: types.django_html = """
+                {% load component_tags %}
+                <div>
+                    {% slot "my_slot" abc=abc 123=var123 %}Default text{% endslot %}
+                </div>
+            """
+
+            def get_context_data(self):
+                return {
+                    "abc": "def",
+                    "var123": 456,
+                }
+
+        template: types.django_html = """
+            {% load component_tags %}
+            {% component "test" %}
+                {% fill "my_slot" %}
+                    overriden
+                    {{ slot_data_in_fill }}
+                {% endfill %}
+            {% endcomponent %}
+        """
+        rendered = Template(template).render(Context())
+        expected = "<div> overriden </div>"
+        self.assertHTMLEqual(rendered, expected)
+
+    def test_slot_data_fill_without_slot_data(self):
+        @component.register("test")
+        class TestComponent(component.Component):
+            template: types.django_html = """
+                {% load component_tags %}
+                <div>
+                    {% slot "my_slot" %}Default text{% endslot %}
+                </div>
+            """
+
+        template: types.django_html = """
+            {% load component_tags %}
+            {% component "test" %}
+                {% fill "my_slot" slot_data="data" %}
+                    {{ data|safe }}
+                {% endfill %}
+            {% endcomponent %}
+        """
+        rendered = Template(template).render(Context())
+        expected = "<div> {} </div>"
```

Comment:
Minor: Where do these brackets come from? Are they empty dict -> str? Shouldn't the default be empty string instead?

Reply:
Yes, it's from doing `{{ data|safe }}`.

The dict is not intended to be printed like I did with `{{ data }}`, I did so only for the sake of tests. So I don't like the idea of coercing empty dict to a string. Instead, it's more likely that the `slot_data` will be used as an object, accessing keys like `{{ data.some_key }}`. So we should keep it as a dict, so users don't get surprised when they try to access a key, and instead they accidentally access a string property/method.